specs/DataSet.rst

```diff
@@ -0,0 +1,253 @@
+   This includes the name, role (set-point or output), and type of each parameter.
+   Parameters may be marked as optional, in which case they are not required for each result.
+#. It should be possible to define a result parameter that is independent of any QCoDeSParameter or Instrument.
+#. A QCoDeS Parameter should provide sufficient information to define a result parameter.
+#. A DataSet should allow storage of relatively arbitrary metadata describing the run that generated the results and the parameters included in the results.
+    
+Writing
+----------
+
+#. It should be possible to add a single result or a sequence of results to an in-progress DataSet.
+#. A DataSet should maintain the order in which results were added.
+#. An in-progress DataSet may be marked as completed.
+
+Access
+---------
+
+#. Values in a DataSet should be easily accessible for plotting and analysis, even while the DataSet is in progress.
+   In particular, it should be possible to retrieve results in a NumPy-compatible form.
+#. It should be possible to define a cursor that specifies a location in a specific value set in a DataSet.
+   It should be possible to get a cursor that specifies the current end of the DataSet when the DataSet is "in progress".
+   It should be possible to read "new data" in a DataSet; that is, to read everything after a cursor.
+#. It should be possible to subscribe to change notifications from a DataSet.
+   It is acceptable if such subscriptions must be in-process until QCoDeS multiprocessing is redone.
+   Change notifications should include the results that were added to the DataSet that triggered the notification.
+
+Storage and Persistence
+-----------------------
+
+#. A DataSet object should allow writing to and reading from storage in a variety of formats.
+#. Users should be able to define new persistence formats.
+#. Users should be able to specify where a DataSet is written.
+
+Interface
+=========
+
+Creation
+--------
+
+ParamSpec
+~~~~~~~~~
+
+A ParamSpec object specifies a single parameter in a DataSet.
+
+ParamSpec(Parameter p, optional=)
+    Creates a parameter specification from a QCoDeS Parameter.
+    If optional is provided and is true, then the parameter is optional in each result.
+    
+ParamSpec(name, role, type, desc=, optional=)
+    Creates a parameter specification with the given name, role (‘I’ or ‘O’), and type. 
+    The type should be a NumPy dtype object.
+    If a description is provided, it is included in the metadata of the DataSet.
+    The description can be a simple string or a string-to-string dictionary.
+    If optional is provided and is true, then the parameter is optional in each result.
+
+DataSet
+~~~~~~~
+
+DataSet()
+    Creates a DataSet with no parameters.
+
+DataSet(specs)
+    Creates a DataSet for the provided list of parameter specifications.
+    Each item in the list should either be a QCoDeS Parameter, a tuple of a Parameter and a Boolean, or a ParamSpec object.
```

Comment:
My impression is that there is a desire to allow certain values (parameters) to be left out of some results, so that's what optional parameters are for. The notion is that add_result should make sure that there's a value for each parameter that's not optional, and fail if anything is missing.

If this checking isn't needed, then we can drop this feature. I'm always happy to drop features :-).